app/controllers/roles_controller.rb

```diff
@@ -0,0 +1,51 @@
+# frozen_string_literal: true
+
+class RolesController < ApplicationController
+  before_action :set_user
+  before_action :set_item, only: [:create]
+
+  def create
+    if @item && @user.has_role?(params[:role], @item)
+      redirect_back(fallback_location: root_path, flash: { alert: "User: #{@user.uid} is already assigned as #{params[:role]}" })
+    elsif @item
+      # each user only gets one role per item, remove all others first.
+      # This removes the role for all users who have that role, not only @user
+      # @user.roles.where(resource: @item).destroy_all
```

Comment:
This is destroying the role for all users who share that role, not just the user we are adding a new role for.